pkg/counter/counter.go

```diff
@@ -12,73 +12,74 @@ type Counter struct {
 	value uint64
 }
 
-// NewCounter creates a new counter with the given initial value.
 func NewCounter(initialValue uint64) Counter {
 	return Counter{value: initialValue}
 }
 
-// Increase increases the counter by the given value (thread-safe).
-func (c *Counter) Increase(values ...uint64) error {
-	_, err := c.IncreaseAndRead(values...)
-	return err
-}
+// Increment
 
-// IncreaseAndRead increases the counter by the given value (thread-safe) and returns the new value.
-func (c *Counter) IncreaseAndRead(values ...uint64) (uint64, error) {
-	if len(values) == 0 {
-		values = append(values, 1)
+// Increment increments the counter by given value (default: 1, thread-safe).
+func (c *Counter) Increment(x ...uint64) error {
```

Comment:
Taking care of corner cases like `x = [0]` or `x = [0,0,...,0]` and sum overflow, I would suggest changing to:

```go
func (c *Counter) Increment(x ...uint64) error {
	val := uint64(0)
	if len(x) == 0 {
		val = 1
	} else {
		for _, v := range x {
			// Check for overflow
			if val > math.MaxUint64-v {
				return errors.New("increment values sum overflow")
			}
			val += v
		}
	}

	if val == 0 {
		return nil
	}

	_, err := c.IncrementValueAndRead(val)
	return err
}
```

Reply:
Just a correction in my suggestion, the sum overflow check should be: `if val+v < val {`.

Reply:
Your suggestion is:

- to short circuit zero values (mm'okay)
- not to continue increasing in case of an overflow (why?) wouldn't it be better to mark how many overflows there was somewhere else instead ? and if we did, i would have to keep track of instanced counters (which is not what we've done so far).

Reply:
How much faster do you think we will be checking for zeros instead of just doing the ops ?

Reply:
My version:

```
goos: linux
goarch: amd64
pkg: github.com/aquasecurity/tracee/pkg/counter
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkIncrement
BenchmarkIncrement-6   	 3241423	       357.1 ns/op	       0 B/op	       0 allocs/op
PASS
coverage: 32.1% of statements
```

Your version:

```
goos: linux
goarch: amd64
pkg: github.com/aquasecurity/tracee/pkg/counter
cpu: 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
BenchmarkIncrement
BenchmarkIncrement-6   	 2338204	       506.2 ns/op	       0 B/op	       0 allocs/op
PASS
coverage: 34.4% of statements
ok  	github.com/aquasecurity/tracee/pkg/counter	1.737s
```

Reply:
>     * not to continue increasing in case of an overflow (why?) wouldn't it be better to mark how many overflows there was somewhere else instead?

If we ignore sum overflow, we masquerade wrong inc/decs on the second stage (Inc/DcrementValueAndRead). Suppose we overflowed (wrapped) the sum val to 1 and ignored the overflow. The current value is 2, adding that 1 from the sum, we got a value not overflowed so not detected as an error. That's why I consider that all stages need to detect possible wrapping and err.

Reply:
You mean Increment(1, Math.MaxUint64) would have 2 overflows and not only 1 ?

Reply:
> How much faster do you think we will be checking for zeros instead of just doing the ops ?

Checking the sum result for zero would avoid the atomic Add and its wrapping check.

Reply:
In theory you are right (about checking wraps twice) but I don't think increases will be made by MaxUint64() in any realistic case, so I don't think its worth checking apart from seeing if the *actual* counter was overflowed.

Reply:
From 357 ns/op -> 500 ns/op by checking if val is zero.
IF testing only Increment(0) in the benchmark loop we do ~141 ns/op.

I don't think its worth the short circuit unless we know the frequency of zeros we're adding in overall tracee execution (orelse we're ALWAYS adding ~100ns/op to something that might not be that frequent).

Reply:
So Id rather have the code I proposed for the reasons specified above.

Reply:
I really don't see current use cases for slice of uints (params). So I would vote to get rid of that API; it's making the counter complex for not real world cases. WDYT?

Reply:
> In theory you are right (about checking wraps twice) but I don't think increases will be made by MaxUint64() in any realistic case, so I don't think its worth checking apart from seeing if the _actual_ counter was overflowed (when we consider benchmarking).

Added a comment in the git log saying that isn't supported (and we are not doing for performance reasons).

Reply:
Funny discovery: 

```go
func BenchmarkDecrement(b *testing.B) {
	c := NewCounter(uint64(b.N))

	for i := b.N; i > 0; i-- {
		err := c.Decrement()
		if err != nil {
			b.Fatal(err)
		}
	}
}

func BenchmarkDecrementWithValue(b *testing.B) {
	c := NewCounter(uint64(b.N))

	for i := b.N; i > 0; i-- {
		err := c.Decrement(1)
		if err != nil {
			b.Fatal(err)
		}
	}
}
```

The benchmark shows that the second is almost 2x slower than the first =) (likely because of stack management, argument passing and calling convention).

Reply:
Even funier: having more than 1 argument is faster than having a single argument (likely because how golang decides to pass the argument: through registers or stack, because of calling convention as well). Crazy.

Reply:
> I really don't see current use cases for slice of uints (params). So I would vote to get rid of that API; it's making the counter complex for not real world cases. WDYT?

I agree (from the beginning), it would make the counter faster actually (not having to deal with a slice each inc/dec).

Reply:
> I agree (from the beginning), it would make the counter faster actually (not having to deal with a slice each inc/dec).

And would have a coherent logic without unlikely corner cases.